accepted/2025/nuget-mcp.md

```diff
@@ -0,0 +1,265 @@
+
+The .NET team is working on a single-shot experience similar to `npx`. This is not work done by the NuGet team, so we'll just link to the existing efforts.
+
+- GitHub issues: [dotnet/sdk#31103](https://github.com/dotnet/sdk/issues/31103), [dotnet/sdk#47517](https://github.com/dotnet/sdk/issues/47517)
+- Design document: [dotnet/designs#334 - Add a design proposal for dotnet tool exec and dnx](https://github.com/dotnet/designs/pull/334)
+  - Proposal PR: [dotnet/sdk#48443](https://github.com/dotnet/sdk/pull/48443)
+- Design document: [dotnet/designs#333 - Add proposal for RID-specific .NET Tool packages](https://github.com/dotnet/designs/pull/333)
+
+The sample JSON above leverages this new single-shot command execution.
+
+RID-specific tools solve the "giant package" problem but are not necessarily a hard blocker for the experience.
+
+### Add support for NuGet packages in the MCP registry
+
+#### Today's experience
+
+The [official MCP registry](https://github.com/modelcontextprotocol/registry/discussions/11) is not yet live, but the discussion and protocol description around it clearly list [several supported underlying registries](https://github.com/modelcontextprotocol/registry/blob/a4cefcf05f81466ad65e7c3971e76d0f6d60783e/docs/openapi.yaml#L173-L176) for hosting the local server's code.
+
+Current list: `npm`, `docker`, `pypi`, `homebrew`
+
+#### Add NuGet to the list
+
+We will work with the MCP server registry team to add `nuget` to the list.
+
+We will publish/link to guidance on how to publish a NuGet-based MCP server to the MCP registry. This step is performed after the package is published to NuGet.org.
+
+### Add support for NuGet packages in VS Code
+
+#### Today's experience
+
+Similar to the previous section, NuGet is not a recognized MCP server host.
+
+![VS Code MCP server installation](../../meta/resources/nuget-mcp/vscode-today.png)
+
+#### Add NuGet to the list
+
+We will work with the VS Code team to add a NuGet option to the list, which updates the user's `mcp.json` to invoke `dnx` with the provided package ID.
+
+This is tracked by [microsoft/vscode-copilot#15329](https://github.com/microsoft/vscode-copilot/issues/15329).
+
+Once Visual Studio has a corresponding experience ([the current experience is manually editing the `mcp.json`](https://learn.microsoft.com/en-us/visualstudio/ide/mcp-servers?view=vs-2022)), we will ensure NuGet is supported similarly.
+
+## Rollout plan
+
+1. Enable `McpServer` support on NuGet.org search and the package details page, behind a feature flag.
+2. Work with Microsoft and community MCP server implementers to package as NuGet.
+   - As of June 2025, there are 28 .NET tools on NuGet.org that use the MCP SDK. Newer versions should have the `McpServer` package type.
+   - Some Microsoft MCP servers are implemented in .NET but distributed via npm, for example [@azure/mcp](https://www.npmjs.com/package/@azure/mcp).
+3. Wait for `dnx` to land in a .NET 10 preview.
```

Comment:
Given `dnx` will be a critical part to this, VS Code and VS should consider what SDK is installed and what `global.json` may be present in the open folder/solution before using it. VS Code itself doesn't guarantee any SDK version is present. VS can guarantee a given .NET SDK is present _if_ the .NET workload is installed, but that SDK may not be available within the open solution if global.json specifies a different (possibly older) SDK version.
If `dnx` isn't available, what should these clients do?

Reply:
I think there are more ways for dnx to fail than those others, because global.json can get in the way. But yes, let's certainly explore how we can predict failure and guide users to resolving it.